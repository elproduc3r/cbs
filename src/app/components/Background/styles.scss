@keyframes fadeIn {
  0%   {  opacity: 0%; }
  100% {  background-size: 100%; }      
 }

.background-gradient {
  transition: background 1s;
  height: 100vh;
  width: 100%;
  max-width: 100%;
  overflow: auto;
  &.animation {
    animation: fadeIn 1s ease-in 1;
  }
  position: absolute;
  top: 0px;
  left: 0px;
  z-index: 1;
}

.background-container {
  height: 100vh;
  width: 100%;
  overflow: auto;
  z-index: 2;
  position: absolute;
}

