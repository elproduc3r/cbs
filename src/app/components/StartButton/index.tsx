"use client"

import { useContext } from "react";
import { BackgroundContext } from "../Background";

const StartButton = () => {

  const {changeBackground} = useContext(BackgroundContext);
  return (
    <button
      onClick={changeBackground}
    >
      Get Started
    </button>
  )
}

export default StartButton;