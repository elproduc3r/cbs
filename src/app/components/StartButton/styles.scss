.start-button {
  padding: 10px 42px;
  background-color: #ffffff;
  border-radius: 28px;
  font-weight: 800;
  font-family: inherit;
  font-size: 1.5rem;
  border-width: 0px;
  align-self: flex-start;
  text-transform: uppercase;
  transition: .5s;
  margin-top: 2rem;
  &:hover {
    color: forestgreen;
  }
}

