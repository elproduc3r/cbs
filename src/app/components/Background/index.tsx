'use client'

import React, {createContext, useState, MouseEventHandler, useEffect} from 'react';
import getRandomColors from '../../utils/getRandomColors';
import classnames from 'classnames';
import './styles.scss';
export interface BackgroundProps {
  children: React.ReactNode;
};

interface BackgroundContextType {
  changeBackground: MouseEventHandler<HTMLButtonElement>;
}

export const BackgroundContext = createContext<BackgroundContextType>({
  changeBackground: () => {}
});

const Background = ({
  children
}: BackgroundProps): JSX.Element => {
  const [colors, setColors] = useState(["#1d6dd5", "#d529cc"]);
  const [animationClassName, setAnimationClassName] = useState('');

  const changeBackground = (): void => {
    const newColors = getRandomColors(2);
    setColors(newColors);
    setAnimationClassName("animation");
  };

  useEffect(() => {
    if(animationClassName === "animation") {
      setTimeout(() => {
        setAnimationClassName("");
      }, 1000);
    }
  }, [animationClassName]);

  /**
   *  Using style obj here instead of styled-component because
   *  it was rendering without any styles. This way SSR rendering
   *  happens with these default styles. Also, I have added all 
   *  styles here so they are all in 1 place and easier to manage.
   */
  const styles = {
    'backgroundImage': `linear-gradient(to bottom right, ${colors[0]}, ${colors[1]})`,
  };

  return (
    <BackgroundContext.Provider value={{changeBackground}}>

        <div className={'background-container'}>
          {children}
        </div>
        <div
          className={classnames('background-gradient', animationClassName)}
          style={styles}
        >
        </div>
    </BackgroundContext.Provider>
  );
};

export default Background;
