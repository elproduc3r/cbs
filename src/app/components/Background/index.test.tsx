import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Background, {BackgroundContext} from '.';
import StartButton from '../StartButton';

test('Background snapshot', () => {
  const {container} = render(
    <Background>
      <div>child text</div>
    </Background>
  );
  expect(container).toMatchSnapshot();
});

test('Background renders children', () => {
  const childText = "child component";
  render(
    <Background>
      <div>{childText}</div>
    </Background>
  );
  const textElement = screen.getByText(childText);
  expect(textElement).toBeInTheDocument();
});

test('Background has a gradient style', () => {
  const childText = "child component";
  render(
    <Background>
      {childText}
    </Background>
  );
  const container = screen.getByText(childText);
  expect(container).toHaveStyle('height: 100vh');
  expect(container).toHaveStyle('background-image: linear gradient');
  expect(container).toHaveStyle('overflow: auto');
});

test('BackgroundContext.Provider function gets called', () => {
  const mockFn = jest.fn(() => {});
  render(
    <BackgroundContext.Provider value={{changeBackground: mockFn}}>
      <button>Click Me</button>
    </BackgroundContext.Provider>
  );
  expect(mockFn).toBeDefined();
});
