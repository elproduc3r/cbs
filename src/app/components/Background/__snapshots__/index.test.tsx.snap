// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Background> snapshot 1`] = `
<div>
  <div
    class="background-container"
  >
    <div>
      child text
    </div>
  </div>
  <div
    class="background-gradient"
  />
</div>
`;
